src/kde/quick_index.hh

```diff
@@ -0,0 +1,146 @@
+//--------------------------------------------*-C++-*---------------------------------------------//
+/*!
+ * \file   kde/quick_index.hh
+ * \author Mathew Cleveland
+ * \brief  This class generates coarse spatial indexing to quickly access
+ * near-neighbor data. This additionally provides simple interpolation schemes
+ * to map data to simple structured meshes. 
+ * \note   Copyright (C) 2018-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//------------------------------------------------------------------------------------------------//
+
+#ifndef rtt_kde_quick_index_hh
+#define rtt_kde_quick_index_hh
+
+#include "c4/global.hh"
+#include <array>
+#include <map>
+#include <vector>
+
+namespace rtt_kde {
+
+//================================================================================================//
+/*!
+ * \class quick_index
+ * \brief
+ *
+ * Provide a hash like index of spatial distributed data along with simple
+ * mapping functions.
+ * 
+ */
+//================================================================================================//
+
+class quick_index {
+public:
+  //! Default constructors.
+  quick_index(const size_t dim, const std::vector<std::array<double, 3>> &locations,
+              const double max_window_size, const size_t bins_per_dimension,
+              const bool domain_decomposed);
+
+  //! Collect Ghost Data
+  void collect_ghost_data(const std::vector<double> &local_data,
+                          std::vector<double> &local_ghost_data) const;
+
+  //! Override function of 3D array ghost data.
+  void collect_ghost_data(const std::vector<std::array<double, 3>> &local_data,
+                          std::vector<std::array<double, 3>> &local_ghost_data) const;
+
+  //! Override function for vector<vector<double> array ghost data.
+  void collect_ghost_data(const std::vector<std::vector<double>> &local_data,
+                          std::vector<std::vector<double>> &local_ghost_data) const;
+
+  //! Fetch list of coarse index values bound by the window
+  std::vector<size_t> window_coarse_index_list(const std::array<double, 3> &window_min,
+                                               const std::array<double, 3> &window_max) const;
+
+  //! Map local+ghost data to grid window
+  void map_data_to_grid_window(const std::vector<double> &local_data,
+                               const std::vector<double> &ghost_data,
+                               std::vector<double> &grid_dataaa,
```

Comment:
I read this like you were yelling. "GRID DATAAAAAA"

Reply:
HAHAHA... Too funny. I guess I got a little aaaaaa haaaappy....